packages/cli/src/register-command.ts

```diff
@@ -0,0 +1,220 @@
+    // An option's name is determined by Commander by removing the leading
+    // dashes and camel-casing the flag. In order to avoid accidental
+    // differences in the algorithm we look it up in the command and use
+    // whatever Commander says.
+    const name: string | undefined = cmd.options.find(o => o.flags === opt.format)?.attributeName();
+
+    if (name === undefined) {
+      // Just a sanity check. Should never happen.
+      throw new Error('Could not find option: ' + opt.format);
+    }
+
+    yield [name, opt];
+  }
+
+  for (const arg of spec.args) {
+    yield [arg.name, arg];
+  }
+}
+
+async function promptForMissing(cmd: Command, spec: CommandSpec, argsAndOpts: ArgsAndOpts) {
+  for (const [name, param] of allParams(cmd, spec)) {
+    const value = argsAndOpts[name];
+    if (value === undefined || (Array.isArray(value) && value.length === 0)) {
+      // Seems redundant but it helps the type system.
+      if (param.variadic === true) {
+        const prompt = await param.prompt?.(argsAndOpts);
+        const values = [];
+        for (const p of prompt) {
+          values.push(await askQuestion(name, p));
+        }
+        argsAndOpts[name] = values;
+      } else {
+        const prompt = await param.prompt?.(argsAndOpts);
+        argsAndOpts[name] = prompt && (await askQuestion(name, prompt));
+      }
+    } else {
+      if (typeof value === 'boolean') {
+        continue;
+      }
+      if (param.variadic === true) {
+        if (!Array.isArray(value)) {
+          throw new Error(`Expected multiple values for ${name}`);
+        }
+        const prompt = await param.prompt?.(argsAndOpts);
+        if (prompt) {
+          if (prompt.length !== value.length) {
+            throw new Error(`Expected ${prompt.length} values for ${name} but got ${value.length}`);
+          }
+
+          for (const [i, p] of prompt.entries()) {
+            if (!p.validate(value[i])) {
+              throw new Error(`Invalid ${name} '${value[i]}'`);
+            }
+          }
+        }
+      } else {
+        if (Array.isArray(value)) {
+          throw new Error(`Expected a single value for ${name}`);
+        }
+        const prompt = await param.prompt?.(argsAndOpts);
+        if (prompt && !prompt.validate(value)) {
+          throw new Error(`Invalid ${name} '${value}'`);
+        }
+      }
+    }
+  }
+}
+
+async function askQuestion(name: string, question: Question): Promise<string> {
+  const { message, choices, validate: rawValidate, validationError } = question;
+  const type = choices === undefined ? 'input' : 'list';
+
+  const validate = (value: string) => {
+    const valid = rawValidate(value);
+    return valid || (validationError ?? `Invalid ${name}`);
+  };
+
+  const answers = await inquirer.prompt({
+    name: 'question',
+    type,
+    message,
+    choices,
+    validate,
+  });
+
+  return answers.question;
+}
+
+export type ArgsAndOpts = Record<string, boolean | string | string[]>;
```

Comment:
Can you push this definition closer to the beginning of the file, since it's used in several places before?